**THE PULSE OF THE CITY **

ACT-1: A Number That Chooses the Story  

In [1]:
import sympy as sp

# 1) Enter last 4 digits
W = 1
X = 2
Y = 3
Z = 4

# 2) Digital root
def digital_root(n):
    while n > 9:
        n = sum(int(d) for d in str(n))
    return n

R = digital_root(W + X + Y + Z)

# 3) Define variable
x = sp.symbols('x')

# 4) Define function f(x)
f = (x**3)/3 - R*(x**2) + (R**2 - 1)*x

# 5) Differentiate
f_prime = sp.diff(f, x)

# 6) Solve f'(x) = 0
critical_points = sp.solve(f_prime, x)

# 7) Output
print("R =", R)
print("f'(x) =", f_prime)
print("Critical Points:", critical_points)

R = 1
f'(x) = x**2 - 2*x
Critical Points: [0, 2]


“The critical points at 0 and 2 represent transition zones where system behavior changes. These define the focused range for further analysis instead of considering the entire dataset.”

ACT-2: The Raw City

In [3]:
import pandas as pd

df = pd.read_csv("yellow_tripdata_2015-01.csv")  # exact name
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,store_and_fwd_flag,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,2,2015-01-15 19:05:39,2015-01-15 19:23:42,1,1.59,-73.993896,40.750111,1,N,-73.974785,40.750618,1,12.0,1.0,0.5,3.25,0.0,0.3,17.05
1,1,2015-01-10 20:33:38,2015-01-10 20:53:28,1,3.30,-74.001648,40.724243,1,N,-73.994415,40.759109,1,14.5,0.5,0.5,2.00,0.0,0.3,17.80
2,1,2015-01-10 20:33:38,2015-01-10 20:43:41,1,1.80,-73.963341,40.802788,1,N,-73.951820,40.824413,2,9.5,0.5,0.5,0.00,0.0,0.3,10.80
3,1,2015-01-10 20:33:39,2015-01-10 20:35:31,1,0.50,-74.009087,40.713818,1,N,-74.004326,40.719986,2,3.5,0.5,0.5,0.00,0.0,0.3,4.80
4,1,2015-01-10 20:33:39,2015-01-10 20:52:58,1,3.00,-73.971176,40.762428,1,N,-74.004181,40.742653,2,15.0,0.5,0.5,0.00,0.0,0.3,16.30


“The dataset contains detailed trip-level information including timestamps, distance, fare, and location. Initial inspection shows that while the structure appears complete, there may exist inconsistencies such as very short trips, zero tip amounts, and potential anomalies in fare and distance values. Further analysis is required to identify invalid or unrealistic trips.”

In [4]:
df.shape
df.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'pickup_longitude',
       'pickup_latitude', 'RateCodeID', 'store_and_fwd_flag',
       'dropoff_longitude', 'dropoff_latitude', 'payment_type', 'fare_amount',
       'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'total_amount'],
      dtype='object')

In [5]:
df.describe()

,VendorID,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06,1.225204e+06
mean,1.522608e+00,1.681475e+00,2.871340e+00,-7.256503e+01,3.997458e+01,1.036508e+00,-7.260611e+01,3.999820e+01,1.386969e+00,1.187181e+01,3.091479e-01,4.977608e-01,1.541555e+00,2.409790e-01,2.834080e-01,1.476046e+01
std,4.994888e-01,1.338358e+00,8.751440e+01,1.011396e+01,5.571684e+00,6.509691e-01,9.967055e+00,5.490774e+00,4.987816e-01,1.017225e+01,3.634982e-01,3.579815e-02,2.435638e+00,1.297986e+00,6.857341e-02,1.236895e+01
min,1.000000e+00,0.000000e+00,0.000000e+00,-8.745187e+01,0.000000e+00,1.000000e+00,-8.959954e+01,0.000000e+00,1.000000e+00,-4.500000e+02,-1.000000e+00,-5.000000e-01,-8.100000e+01,-1.175000e+01,0.000000e+00,-4.503000e+02
25%,1.000000e+00,1.000000e+00,1.000000e+00,-7.399168e+01,4.073569e+01,1.000000e+00,-7.399123e+01,4.073443e+01,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,3.000000e-01,8.160000e+00
50%,2.000000e+00,1.000000e+00,1.660000e+00,-7.398158e+01,4.075326e+01,1.000000e+00,-7.397979e+01,4.075368e+01,1.000000e+00,9.000000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,3.000000e-01,1.115000e+01
75%,2.000000e+00,2.000000e+00,3.000000e+00,-7.396663e+01,4.076768e+01,1.000000e+00,-7.396250e+01,4.076890e+01,2.000000e+00,1.350000e+01,5.000000e-01,5.000000e-01,2.060000e+00,0.000000e+00,3.000000e-01,1.630000e+01
max,2.000000e+00,9.000000e+00,9.200090e+04,0.000000e+00,4.938905e+01,9.900000e+01,0.000000e+00,5.747092e+01,4.000000e+00,9.499900e+02,7.000000e+00,5.000000e-01,7.500000e+02,3.153300e+02,3.000000e-01,1.501290e+03


“The dataset contains invalid values such as zero or negative distances, fares, and passenger counts, along with extreme outliers. High variability and inconsistent values indicate the presence of erroneous records, making data cleaning necessary.”

In [6]:
df.isnull().sum()

,0
VendorID,0
tpep_pickup_datetime,0
tpep_dropoff_datetime,0
passenger_count,0
trip_distance,0
pickup_longitude,0
pickup_latitude,0
RateCodeID,0
store_and_fwd_flag,0
dropoff_longitude,0


“No significant missing values were observed in the dataset

In [7]:
df_invalid = df[
    (df['trip_distance'] <= 0) |
    (df['fare_amount'] <= 0) |
    (df['passenger_count'] <= 0)
]

df_invalid.shape

(9013, 19)

“A total of 9013 records were identified as invalid based on conditions such as non-positive trip distance, fare amount, or passenger count. This confirms the presence of unrealistic trips within the dataset that do not represent actual taxi activity.”

In [9]:
df_clean = df[
    (df['trip_distance'] > 0) &
    (df['fare_amount'] > 0) &
    (df['passenger_count'] > 0) &
    (df['tpep_dropoff_datetime'] > df['tpep_pickup_datetime'])
]

In [10]:
print("Original:", len(df))
print("Cleaned:", len(df_clean))

Original: 1225204
Cleaned: 1216106


“Although the number of removed records is relatively small compared to the total dataset, their presence can still impact analysis, especially when identifying patterns or anomalies.                                      “Although the dataset contains no missing values, the presence of invalid and unrealistic records highlights that completeness does not guarantee data quality. Careful validation is therefore essential before analysis.”

“Even a small proportion of invalid records can significantly affect aggregated metrics such as average fare or trip distance, potentially leading to misleading conclusions.”

ACT-3: The People Inside the System

---

Stating hypothesis and Statistical Reasoning

* “Passengers who pay higher fares tend to give higher tips.”



In [11]:
df_clean[['fare_amount', 'tip_amount']].corr()

,fare_amount,tip_amount
fare_amount,1.000000,0.562401
tip_amount,0.562401,1.000000


“A correlation value of 0.562 was observed between fare amount and tip amount, indicating a moderate positive relationship. This suggests that as the fare increases, the tip amount also tends to increase.”

**ALTERNATIVE EXPLAINATION**

---

“However, this relationship may not purely reflect passenger behavior. Higher fares are often associated with longer trip distances, which naturally result in higher tip amounts even if passengers follow a fixed tipping rate.”

“At this stage, the alternative explanation appears equally strong, as the data does not distinguish between absolute tip values and proportional tipping behavior.”

**CONDITIONS THAT COULD CHANGE**

---

“This conclusion could be challenged if further analysis reveals that tip percentage remains constant across different fare levels. Additionally, factors such as payment method, time of day, or trip purpose may influence tipping behavior and alter the observed pattern.”

ACT-4: The City as a Signal

In [15]:
# convert datetime
df_clean.loc[:, 'tpep_pickup_datetime'] = pd.to_datetime(df_clean['tpep_pickup_datetime'])

In [16]:
df_clean['hour'] = df_clean['tpep_pickup_datetime'].dt.hour
df_clean['day'] = df_clean['tpep_pickup_datetime'].dt.day_name()

/tmp/ipykernel_893/2600195348.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['hour'] = df_clean['tpep_pickup_datetime'].dt.hour
/tmp/ipykernel_893/2600195348.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['day'] = df_clean['tpep_pickup_datetime'].dt.day_name()


In [18]:
#trip volume by hour
hourly_trips = df_clean.groupby('hour').size()
hourly_trips

,0
hour,
0,46658
1,31114
2,24191
3,17339
4,10842
5,12667
6,22587
7,46482
8,54785


In [19]:
daily_trips = df_clean.groupby('day').size()
daily_trips

,0
day,
Friday,205621
Monday,123569
Saturday,234705
Sunday,144151
Thursday,207672
Tuesday,137102
Wednesday,163286


“Daily aggregation of trip volume shows clear variation across days. Trip activity is highest on Saturdays, followed by Thursdays and Fridays, while Mondays show the lowest activity. Mid-week days exhibit a gradual increase in trips.”

**Which patterns persist**

“The increase in activity towards the end of the week and the peak on weekends appear consistently, indicating a stable and recurring pattern.”

**Which patterns disappear**

“Minor variations between adjacent weekdays are not consistent and likely represent random fluctuations rather than meaningful patterns.”

which were never there to begin with

“The absence of consistent patterns at the individual trip level, combined with the emergence of stable trends after aggregation, confirms that earlier variations were random noise rather than meaningful system behavior.”

“The city’s rhythm is cyclical rather than random noise. This is confirmed by the consistent and repeating patterns observed in the aggregated data. Trip volumes show predictable peaks during evening hours and weekends, and consistent drops during early morning periods. These patterns persist across different levels of aggregation, indicating that they are driven by structured human behavior such as commuting and social activity.

Since these trends repeat regularly and do not disappear upon aggregation, they cannot be considered noise. Therefore, the system exhibits a stable and cyclical rhythm rather than random fluctuations that only appear convincing.”

**ACT-5: Pressure on the System**

“Fare amount is selected as the key parameter because it directly influences passenger behavior, revenue, and tipping patterns, making it a critical factor in the system.”

In [20]:
#Applying +1% change as the R value considered in ACT-1 is = 1
R = 1

df_sim = df_clean.copy()

# Increase fare by R%
df_sim['fare_amount_new'] = df_sim['fare_amount'] * (1 + R/100)

df_sim[['fare_amount', 'fare_amount_new']].head()

,fare_amount,fare_amount_new
0,12.0,12.120
1,14.5,14.645
2,9.5,9.595
3,3.5,3.535
4,15.0,15.150


“A 1% increase was applied to the fare amount, resulting in a proportional increase in all fare values. The transformation is consistent across all records, confirming that the simulation has been correctly implemented.”

In [21]:
df_clean['fare_amount'].mean(), df_sim['fare_amount_new'].mean()

(np.float64(11.836675610514218), np.float64(11.955042366619363))

In [22]:
df_clean['fare_amount'].sum(), df_sim['fare_amount_new'].sum()

(np.float64(14394652.230000004), np.float64(14538598.752300007))

“The average fare increased slightly after applying the 1% change, indicating a proportional adjustment across all trips. The small magnitude of change suggests that the overall pricing structure of the system remains stable and is not significantly disrupted.”


“The total revenue shows a noticeable increase following the fare adjustment, confirming that even a small percentage change can lead to a significant cumulative impact at scale. This highlights the sensitivity of total system revenue to pricing changes.”


“The system absorbs the change in terms of behavior, but amplifies it in terms of aggregate revenue.”

In [24]:
df_clean.to_csv("cleaned_taxi_data.csv", index=False)

In [25]:
from google.colab import files
files.download("cleaned_taxi_data.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>